
# **What is Python?**

Python is an interpreted, object-oriented, high-level programming language with dynamic semantics. Its high-level built-in data structures, combined with dynamic typing and dynamic binding, make it very attractive for Rapid Application Development as well as for use as a scripting or glue language to connect existing components together.Python is simple and easier to learn. It puts emphasis on readability, and thereby lowers the cost of program maintenance. Python supports modules and packages, which allows for program modularity and code reuse. The Python interpreter and the vast standard library are freely available in source or binary form for all major platforms and can be freely distributed.

# **History of Python**

Python was created in the early 1990s by Guido van Rossum at Stichting Mathematisch Centrum (CWI) in the Netherlands as a successor of a language called ABC. Guido remains Python’s principal author, although it includes many contributions from others.

In 1995, Guido continued his work on Python at the Corporation for National Research Initiatives (CNRI) in Reston, Virginia where he released several versions of the software.

In May 2000, Guido and the Python core development team moved to BeOpen.com to form the BeOpen PythonLabs team. In October of the same year, the PythonLabs team moved to Digital Creations (now Zope Corporation).

In 2001, the Python Software Foundation (PSF) was formed, a non-profit organization created specifically to own Python-related Intellectual Property. Zope Corporation is a sponsoring member of the PSF.

All Python releases are Open Source. Historically, most, but not all, Python releases have also been GPL-compatible; the table below summarizes the various releases.

# **Features of Python**

**Easy to Learn and Use**: Python’s simple syntax and readability make it ideal for beginners.

**Expressive Language** : Python allows complex tasks to be performed with minimal code.

**Interpreted Language**: Python executes code line-by-line, facilitating debugging and portability.

**Cross-platform Languag**e: Python runs on various platforms without modification, enhancing its portability.

**Free and Open Source**: Python is freely available and open for contributions from a global community.

**Object-Oriented Language**: Python supports object-oriented programming with features like inheritance and encapsulation.

**Extensible**: Python can interface with other languages like C/C++ to extend its functionality.

**Large Standard Library**: Python offers a comprehensive library collection for diverse applications, from web development to machine learning.

**GUI Programming Support**: Python provides libraries like Tkinter and PyQt for developing desktop applications with graphical interfaces.

**Integrated**: Python integrates seamlessly with other languages, making it versatile and easy to debug.

**Embeddable**: Python code can be embedded within other languages and vice versa.

**Dynamic Memory Allocation**: Python automatically manages memory allocation for variables at runtime without explicit type declarations.

# **Applications of Python**

Web Applications: Python is used to build web applications with frameworks that handle various internet protocols and content management.
Examples: Django (used by Instagram), Flask (lightweight web framework), Plone (content management).
 2. Desktop GUI Applications: Python supports GUI development, allowing the creation of desktop applications with interactive interfaces.

Examples: Tkinter (standard GUI library), PyQt (advanced GUI toolkit), Kivy (multitouch applications).
 3. Console-based Applications: Python can develop command-line applications, leveraging its REPL environment for interactive programming.

Examples: Command-line tools using libraries like argparse, and interactive scripts for various tasks.
 4. Software Development: Python aids in software development for tasks like build control, automated testing, and project management.

Examples: SCons (build control), Buildbot (automated testing), Trac (project management).
 5. Scientific and Numeric: Python excels in scientific and numeric computing with libraries for complex mathematical and data analysis tasks.

Examples: NumPy (numerical operations), SciPy (scientific computing), Pandas (data manipulation).
 6. Business Applications: Python supports business applications requiring scalability and integration, such as e-commerce and ERP systems.

Examples: Odoo (ERP system), Tryton (business application platform).
 7. Audio or Video-based Applications: Python can be used to develop multimedia applications, handling audio and video processing tasks.

Examples: Gstreamer (multimedia framework), Pyglet (video and audio library).
 8. 3D CAD Applications: Python facilitates the creation of 3D CAD applications for engineering and architectural design.

Examples: FreeCAD (popular open-source CAD), HeeksCNC (CNC applications).
 9. Enterprise Applications: Python is used for building enterprise-level applications, enhancing organizational efficiency.

Examples: OpenERP (enterprise resource planning), Picalo (business management).
 10. Image Processing Applications: Python offers libraries for image manipulation and processing tasks, enabling custom image processing solutions.

Examples: OpenCV (image processing library) , Pillow (PIL fork for image handling).

# **Advantages of Python**

1.Presence of third-party modules: Python has a rich ecosystem of third-party modules and libraries that extend its functionality for various tasks.

2.Extensive support libraries: Python boasts extensive support libraries like NumPy for numerical calculations and Pandas for data analytics, making it suitable for scientific and data-related applications.

3.Open source and large active community base: Python is open source, and it has a large and active community that contributes to its development and provides support.

4.Versatile, easy to read, learn, and write: Python is known for its simplicity and readability, making it an excellent choice for both beginners and experienced programmers.

5.User-friendly data structures: Python offers intuitive and easy-to-use data structures, simplifying data manipulation and management.

6.High-level language: Python is a high-level language that abstracts low-level details, making it more user-friendly.

7.Dynamically typed language: Python is dynamically typed, meaning you don’t need to declare data types explicitly, making it flexible but still reliable.

8.Object-Oriented and Procedural programming language: Python supports both object-oriented and procedural programming, providing versatility in coding styles.

9.Portable and interactive: Python is portable across operating systems and interactive, allowing real-time code execution and testing. Ideal for prototypes: Python’s concise syntax allows developers to prototype applications quickly with less code.

# **Disadvantages of Python**

1.Performance: Python is an interpreted language, which means that it can be slower than compiled languages like C or Java. This can be an issue for performance-intensive tasks.

2.Global Interpreter Lock: The Global Interpreter Lock (GIL) is a mechanism in Python that prevents multiple threads from executing Python code at once. This can limit the parallelism and concurrency of some applications.

3.Memory consumption: Python can consume a lot of memory, especially when working with large datasets or running complex algorithms.

4.Dynamically typed: Python is a dynamically typed language, which means that the types of variables can change at runtime. This can make it more difficult to catch errors and can lead to bugs.

5.Packaging and versioning: Python has a large number of packages and libraries, which can sometimes lead to versioning issues and package conflicts.

6.Lack of strictness: Python’s flexibility can sometimes be a double-edged sword. While it can be great for rapid development and prototyping, it can also lead to code that is difficult to read and maintain.

7.Steep learning curve: While Python is generally considered to be a relatively easy language to learn, it can still have a steep learning curve for beginners, especially if they have no prior experience with programming.

**Python First Program**

In [1]:
# This program prints Hello, world!

print('Hello, world!')

Hello, world!


# **Comments in Python**

**Single Line Comments**

Single-line remarks in Python have shown to be effective for providing quick descriptions for parameters, function definitions, and expressions. A single-line comment of Python is the one that has a hashtag # at the beginning of it and continues until the finish of the line. If the comment continues to the next line, add a hashtag to the subsequent line and resume the conversation.

In [2]:
# This code is to show an example of a single-line comment
print( 'This statement does not have a hashtag before it' )

This statement does not have a hashtag before it


In [3]:
# This code is to show an example of a single-line comment

The Python compiler ignores this line.

**Multi Line Comments**

Python does not provide the facility for multi-line comments. However, there are indeed many ways to create multi-line comments.In Python, we may use hashtags (#) multiple times to construct multiple lines of comments. Every line with a (#) before it will be regarded as a single-line comment.

In [4]:
# it is a
# comment
# extending to multiple lines

In this case, each line is considered a comment, and they are all omitted.

# **Keywords in Python**

Python keywords are special reserved words that have specific meanings and purposes and can’t be used for anything but those specific purposes. These keywords are always available—you’ll never have to import them into your code.

Python keywords are different from Python’s built-in functions and types. The built-in functions and types are also always available, but they aren’t as restrictive as the keywords in their usage.

An example of something you can’t do with Python keywords is assign something to them. If you try, then you’ll get a SyntaxError. You won’t get a SyntaxError if you try to assign something to a built-in function or type, but it still isn’t a good idea. For a more in-depth explanation of ways keywords can be misused, check out Invalid Syntax in Python: Common Reasons for SyntaxError.

As of Python 3.8, there are thirty-five keywords in Python.

False - Boolean value indicating false.

None - Represents the absence of a value or a null value.

True - Boolean value indicating true.

and - Logical AND operator.

as - Used to create an alias while importing a module.

assert - Used for debugging to test a condition.

break - Exits the nearest enclosing loop.

class - Used to define a new class.

continue - Continues to the next iteration of the nearest enclosing loop.

def - Used to define a function.

del - Deletes a reference to an object.

elif - Used in conditional statements; stands for "else if".

else - Executes a block of code if no preceding conditions are met.

except - Catches exceptions raised by try.

finally - Ensures code executes regardless of whether an exception occurred.

for - Creates a for loop.

from - Used to import specific parts of a module.

global - Declares a global variable accessible across functions.

if - Initiates a conditional statement.

import - Imports a module or its components.

in - Tests membership in an iterable or iterates over an iterable.

is - Tests object identity.

lambda - Creates an anonymous function.

nonlocal - Refers to a variable in an enclosing scope that is not global.

not - Logical NOT operator.

or - Logical OR operator.

pass - A null statement used as a placeholder.

raise - Raises an exception.

return - Exits a function and optionally returns a value.

try - Starts a block of code that will be tested for exceptions.

while - Creates a while loop.

with - Wraps the execution of a block with methods defined by a context manager.

yield - Produces a generator in a function.

walrus - This refers to the "walrus operator" (:=), which allows assignment expressions.

match - This is part of the new pattern matching feature introduced in Python 3.10, but it is reserved in Python 3.8 for future use.

In [5]:
import keyword

py_keywords=keyword.kwlist

print(py_keywords)

len(py_keywords)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


35

Channels:
 - conda-forge
 - defaults
Platform: osx-arm64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /Users/ruchithreddyparnem/opt/anaconda3

  added / updated specs:
    - pandoc


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2024.8.30  |       hf0a4a13_0         155 KB  conda-forge
    certifi-2024.7.4           |     pyhd8ed1ab_0         156 KB  conda-forge
    conda-24.7.1               |  py312h81bd7bf_0         1.2 MB  conda-forge
    libexpat-2.6.2             |       hebf3989_0          62 KB  conda-forge
    libsqlite-3.46.0           |       hfb93653_0         811 KB  conda-forge
    libzlib-1.2.13             |       hfb2fe0b_6          46 KB  conda-forge
    openssl-3.3.1              |       h8359307_3         2.8 MB  conda-forge
    pandoc-3.3                 |       hce30654_0        21.8 MB  conda-forge
    python-3.12.2              |hdf0ec26_0_cpython        12.5 MB  conda-forge
    python_abi-3.12            |          5_cp312           6 KB  conda-forge
    zlib-1.2.13                |       hfb2fe0b_6          76 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        39.5 MB

The following NEW packages will be INSTALLED:

  libexpat           conda-forge/osx-arm64::libexpat-2.6.2-hebf3989_0

  libsqlite          conda-forge/osx-arm64::libsqlite-3.46.0-hfb93653_0

  libzlib            conda-forge/osx-arm64::libzlib-1.2.13-hfb2fe0b_6

  pandoc             conda-forge/osx-arm64::pandoc-3.3-hce30654_0

  python_abi         conda-forge/osx-arm64::python_abi-3.12-5_cp312

The following packages will be UPDATED:

  ca-certificates    pkgs/main::ca-certificates-2024.7.2-h~ -->

  conda-forge::ca-certificates-2024.8.30-hf0a4a13_0

  openssl              pkgs/main::openssl-3.0.14-h80987f9_0 -->
  conda-forge::openssl-3.3.1-h8359307_3

  zlib                    pkgs/main::zlib-1.2.13-h18a0788_1 --> conda-forge::zlib-1.2.13-hfb2fe0b_6


The following packages will be SUPERSEDED by a higher-priority channel:

  certifi            pkgs/main/osx-arm64::certifi-2024.7.4~ --> conda-forge/noarch::certifi-2024.7.4-pyhd8ed1ab_0

  conda              pkgs/main::conda-24.7.1-py312hca03da5~ --> conda-forge::conda-24.7.1-py312h81bd7bf_0

  python                pkgs/main::python-3.12.4-h99e199e_1 --> conda-forge::python-3.12.2-hdf0ec26_0_cpython



Downloading and Extracting Packages:

pandoc-3.3           | 21.8 MB   |                                       |   0%

python-3.12.2        | 12.5 MB   |                                       |   0%

openssl-3.3.1        | 2.8 MB    |                                       |   0%


conda-24.7.1         | 1.2 MB    |                                       |   0%



libsqlite-3.46.0     | 811 KB    |                                       |   0%




certifi-2024.7.4     | 156 KB    |                                       |   0%





ca-certificates-2024 | 155 KB    |                                       |   0%






zlib-1.2.13          | 76 KB     |                                       |   0%







libexpat-2.6.2       | 62 KB     |                                       |   0%








libzlib-1.2.13       | 46 KB     |                                       |   0%









pandoc-3.3           | 21.8 MB   |                                       |   0%

python-3.12.2        | 12.5 MB   |                                       |   0%


conda-24.7.1         | 1.2 MB    | 4                                     |   1%



pandoc-3.3           | 21.8 MB   |                                       |   0%


conda-24.7.1         | 1.2 MB    | #######4                              |  20%

python-3.12.2        | 12.5 MB   | 7                                     |   2%



libsqlite-3.46.0     | 811 KB    | #####8                                |  16%


pandoc-3.3           | 21.8 MB   | 2                                     |   1%

python-3.12.2        | 12.5 MB   | #7                                    |   5%


conda-24.7.1         | 1.2 MB    | ##################3                   |  50%



libsqlite-3.46.0     | 811 KB    | ###########6                          |  32%

python-3.12.2        | 12.5 MB   | ##3                                   |   6%



libsqlite-3.46.0     | 811 KB    | #########################5            |  69%


pandoc-3.3           | 21.8 MB   | 3                                     |   1%

python-3.12.2        | 12.5 MB   | ##8                                   |   8%


conda-24.7.1         | 1.2 MB    | ############################2         |  76%



libsqlite-3.46.0     | 811 KB    | ################################8     |  89%

python-3.12.2        | 12.5 MB   | ###4                                  |   9%



libsqlite-3.46.0     | 811 KB    | ##################################### | 100%


pandoc-3.3           | 21.8 MB   | 6                                     |   2%

python-3.12.2        | 12.5 MB   | ####2                                 |  11%




pandoc-3.3           | 21.8 MB   | 7                                     |   2%




certifi-2024.7.4     | 156 KB    | ##################################### | 100%


conda-24.7.1         | 1.2 MB    | ##################################### | 100%





ca-certificates-2024 | 155 KB    | ###8                                  |  10%

python-3.12.2        | 12.5 MB   | ####8                                 |  13%





ca-certificates-2024 | 155 KB    | ##################################### | 100%

pandoc-3.3           | 21.8 MB   | #4                                    |   4%







libexpat-2.6.2       | 62 KB     | #########5                            |  26%







libexpat-2.6.2       | 62 KB     | ##################################### | 100%






zlib-1.2.13          | 76 KB     | #######7                              |  21%

python-3.12.2        | 12.5 MB   | ######4                               |  18%






pandoc-3.3           | 21.8 MB   | #8                                    |   5%








libzlib-1.2.13       | 46 KB     | ############9                         |  35%








libzlib-1.2.13       | 46 KB     | ##################################### | 100%









python_abi-3.12      | 6 KB      | ##################################### | 100%









python_abi-3.12      | 6 KB      | ##################################### | 100%

python-3.12.2        | 12.5 MB   | #######2                              |  20%

pandoc-3.3           | 21.8 MB   | ##1                                   |   6%

openssl-3.3.1        | 2.8 MB    | ##3                                   |   6%

pandoc-3.3           | 21.8 MB   | ##5                                   |   7%

openssl-3.3.1        | 2.8 MB    | #####2                                |  14%

python-3.12.2        | 12.5 MB   | ########8                             |  24%

pandoc-3.3           | 21.8 MB   | ##8                                   |   8%

pandoc-3.3           | 21.8 MB   | ###1                                  |   9%

python-3.12.2        | 12.5 MB   | ##########                            |  27%

openssl-3.3.1        | 2.8 MB    | ###########3                          |  31%

openssl-3.3.1        | 2.8 MB    | ######################2               |  60%

python-3.12.2        | 12.5 MB   | ##########7                           |  29%

pandoc-3.3           | 21.8 MB   | ###4                                  |   9%

python-3.12.2        | 12.5 MB   | ###########2                          |  30%

pandoc-3.3           | 21.8 MB   | ###8                                  |  10%

pandoc-3.3           | 21.8 MB   | ####1                                 |  11%

openssl-3.3.1        | 2.8 MB    | ##################################### | 100%

openssl-3.3.1        | 2.8 MB    | ##################################### | 100%

pandoc-3.3           | 21.8 MB   | ####7                                 |  13%

pandoc-3.3           | 21.8 MB   | #####2                                |  14%

pandoc-3.3           | 21.8 MB   | #####5                                |  15%

pandoc-3.3           | 21.8 MB   | #####9                                |  16%

python-3.12.2        | 12.5 MB   | ################7                     |  45%

pandoc-3.3           | 21.8 MB   | ######4                               |  17%

pandoc-3.3           | 21.8 MB   | ######9                               |  19%

pandoc-3.3           | 21.8 MB   | #######3                              |  20%

pandoc-3.3           | 21.8 MB   | #######7                              |  21%

pandoc-3.3           | 21.8 MB   | ########1                             |  22%

pandoc-3.3           | 21.8 MB   | ########5                             |  23%

pandoc-3.3           | 21.8 MB   | ########8                             |  24%

pandoc-3.3           | 21.8 MB   | #########2                            |  25%

python-3.12.2        | 12.5 MB   | ############################3         |  77%

pandoc-3.3           | 21.8 MB   | ##########1                           |  27%

pandoc-3.3           | 21.8 MB   | ##########9                           |  30%

pandoc-3.3           | 21.8 MB   | ###########3                          |  31%

pandoc-3.3           | 21.8 MB   | #############6                        |  37%

python-3.12.2        | 12.5 MB   | ##################################### | 100%
                                                                                
                                                                                

                                                                                


                                                                                



                                                                                




                                                                                





                                                                                






                                                                                







                                                                                








                                                                                









                                                                                
Preparing transaction: done

Verifying transaction: done

Executing transaction: done

Note: you may need to restart the kernel to use updated packages.